File: docs/leetcode/40/40.py
# -*- coding:utf-8 -*-

from typing import List 

class Solution:
    def dfs(self, start, target, candidates, line, ans):
        if target == 0:
            print(line)
            ans.append(line[:])
            return
        if start == len(candidates):
            return
        for i in range(start, len(candidates)):
            if i > start and candidates[i] == candidates[i - 1]:
                continue
            if candidates[i] > target:
                break
            line.append(candidates[i])
            self.dfs(i + 1, target - candidates[i], candidates, line, ans)
            line.pop()
        return

    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        ans = []
        if len(candidates) == 0:
            return ans
        
        candidates = sorted(candidates)
        
        line = []
        for i in range(len(candidates)):
            if i > 0 and candidates[i] == candidates[i - 1]:
                continue
            if candidates[i] > target:
                break
            line.append(candidates[i])
            self.dfs(i + 1, target - candidates[i], candidates, line, ans)
            line.pop()
        return ans

s = Solution()
ans = s.combinationSum2([10,1,2,7,6,1,5], 8)
print(ans)

File: docs/leetcode/40/40.cpp
#include <vector>
#include <string>
#include <stack>
#include <algorithm>
#include <deque>

using namespace std; 

class Solution {
public:
    void dfs(int start, int target, vector<int>& candidates, vector<int> &line, vector<vector<int>> &ans) {
        if (target == 0) {
            ans.push_back(line);
            return;
        }
        if (start == candidates.size()) {
            return;
        }
        for (int i = start; i < candidates.size(); i++) {
            if (i > start && candidates[i] == candidates[i-1]) {
                continue;
            }
            if (candidates[i] > target) {
                break;
            }
            line.push_back(candidates[i]);
            dfs(i + 1, target - candidates[i], candidates, line, ans);
            line.pop_back();
        }
        return;
    }
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        vector<vector<int>> ans;

        if (candidates.empty()) {
            return ans;
        }

        sort(candidates.begin(), candidates.end());

        vector<int> line;
        for (int i = 0; i < candidates.size(); i++) {
            if (i > 0 && candidates[i] == candidates[i-1]) {
                continue;
            }
            if (candidates[i] > target) {
                break;
            }
            line.push_back(candidates[i]);
            dfs(i + 1, target - candidates[i], candidates, line, ans);
            line.pop_back();
        }

        return ans;
    }
};

int main()
{
    Solution *s = new Solution();
    vector<int> candidates = {10,1,2,7,6,1,5};
    int target = 8;
    auto ans = s->combinationSum2(candidates, target);
    return 0;
}